feat(db_query)!: use query builder

akhilnarang opened the pull request:
TL;DR: Replaces internal implementation of `get_all`/`get_list` with QB based `get_query`.

---

`qb_query.py` + `class DatabaseQuery` was created with an API as close to `db_query.py` as possible, besides a few cases where we had to break compatibility.

Migration guide: https://github.com/frappe/frappe/wiki/query-builder-migration

Added a mention, updated an example: https://docs.frappe.io/framework/user/en/api/database

<hr>

Resolves #34131

akhilnarang committed:
feat: add in initial version of DatabaseQuery using query builder

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor: use qb_query for get_list

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor: support new function style

- Migrate all SQL function usage from string format to dict format
- Old: fields=['count(*) as count']
- New: fields=[{'COUNT': '*', 'as': 'count'}]
- Add `NULLIF`

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor: backticks aren't allowed in order_by or group_by

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(search): adjust query formation to align with new restrictions

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor: migrate a few queries to use query builder

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(query_builder): default sorting based on doctype meta

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: support virtual doctype handling

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(query): allow `AggregateFunction` as well in `apply_field_permissions`

Without this `fields=[{"COUNT": "name"}]` didn't work, although fields=[{"COUNT": "NAME"}] did.

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(query): allow passing `as` in any case

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(query): allow numeric strings

For things like `COUNT(1)`

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(query): match db_query, add parentheses around condition

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(test_server_script): loosen test a bit, allow any case

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(convert_to_value): convert dict_keys and dict_values to a tuple as well

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: make sure limit is an integer

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(linked_with): don't add `tab` prefix

query builder takes care of this

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat(query): add in IFNULL logic from db_query

Use `IFNULL(var, "") == ""` instead of `isnull()`

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: allow function aliases

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: function detection

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: adjust tests for query builder permission

Earlier you could use fields you didn't have access to for filtering, now you can't.

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor: uppercase function assertions

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(reportview): temporarily skip validation for dict fields

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: tests

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(query): check whether filter fields belong to child tables if not part of parent

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

ankush commented on frappe/core/doctype/report/report.py:
```diff
@@ -291,7 +291,7 @@ def run_standard_report(self, filters, limit, user):
 	def _format(parts):
 		# sort by is saved as DocType.fieldname, covert it to sql
-		return "`tab{}`.`{}`".format(*parts)
+		return parts[1]
```
Use field here. Doctype might be important.

ankush commented on frappe/core/doctype/data_export/exporter.py:
```diff
@@ -370,11 +370,12 @@ def add_data(self):
-		self.data = frappe.get_list(
-			self.doctype, fields=["*"], filters=self.filters, limit_page_length=None, order_by=order_by
-		)
+		self.data = frappe.qb.get_query(
```
Should try to make this work directly with get_list

ankush commented on frappe/core/doctype/user_type/user_type.py:
```diff
@@ -234,7 +234,7 @@ def get_user_linked_doctypes(doctype, txt, searchfield, start, page_len, filters
 		"DocType",
 		fields=["`tabDocType`.`name`"],
 		filters=filters,
-		order_by="`tabDocType`.`idx` desc",
```
We can probably just implement this in qb_query.py parse the expression and convert it into `DocType(dt_name).field`

ankush commented:
Let's try to get this merged, we can later add support for parsing:
- backtick + table + field expression in order_by 
- popular functions

This way, there won't be much migration pain and we still get the benefit of having a smaller surface area. We can migrate our apps fully in the meantime.

ankush commented on frappe/tests/test_db.py:
```diff
@@ -132,7 +132,7 @@ def test_get_value(self):
 		delimiter = '"' if frappe.db.db_type == "postgres" else "`"
 		self.assertIn(
-			"ORDER BY {deli}creation{deli} DESC,{deli}modified{deli} ASC,{deli}name{deli} DESC".format(
+			"ORDER BY {deli}creation{deli} DESC,{deli}modified{deli} ASC,{deli}name{deli} ASC".format(
```
This seems problematic. `get_value` should retain old behaviour, used a lot in business logic to find last record.

akhilnarang committed:
refactor: ensure no meta recursion

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat: support certain backticked expressions

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: check for shared documents

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(check_field_permissions): allow if no permissions defined

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(list): fix `distinct` usage

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(query): case insensitive check

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat: add in some functions

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: skip group by permission check for certain columns

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: test

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: match db_query behaviour for certain cases like

`filters.append(["reports_to", "=", ""])`

Earlier this generated:
```
( `tabEmployee`.`reports_to` is NULL OR `tabEmployee`.`reports_to` = '' )
```

Without this change, with qb it was

```
`reports_to`=''
```

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(query): ensure backwards compatibility for sorting, filtering

If `db_query_compat=True` (set by `qb_query.py`), then we default to some `db_query.py` behaviour.
Otherwise, we'll retail the previous query builder behaviour, this is to minimize breakage on either side.

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(query): extend regex for allow backticked aliases

For example:

```
`tabSerial and Batch Entry`.`name` as `child_row`
```

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: handle converting datetime -> date for fieldtype date

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

ankush commented on frappe/model/meta.py:
```diff
@@ -101,6 +101,10 @@ def clear_meta_cache(doctype: str = "*"):
+def clear_doctype_sort_cache(doctype: str):
+	frappe.cache.delete_value(f"doctype_sort_info::{doctype}")
+
+
```
```suggestion
```

Not required anymore

ankush committed:
chore: remove dead code

ankush commented:
Moved TODOs to a separate issue: https://github.com/frappe/frappe/issues/34772
